### DISCUSSION TITLE
REPE RPC event support

### DISCUSSION THREAD

**Discussion by SiebrenW** [stephenberry/glaze#986] [General]
Hi.

I have gotten interested in REPE as an RPC protocol. I gather there is excellent command support (request, response from client to server and back), but was thinking that not all communication goes one way, though it can be made to work like this, it is often not ideal to poll until the result is what you need from the client side.

For this an RPC protocol would sometimes have some event or state representation that a clients can subscribe to and at any given time receive an event from the server.

I can think of a few ways to make this work, like keep sending a command and once you get a response (the event) you "resubscribe" by sending another. Or host a server on the client too (though this will cause issues with port forwarding and the like) that the server sends these "events" as a request etc. 

Both are not ideal, so I'm hoping that I'm just missing this feature.

So is this already possible in REPE/Glaze? If so; how?

**Comment by stephenberry** [comment]
Thanks for your interest in REPE. It's still highly under development so I love discussing topics like this. I won't have time until next week to consider your question carefully, but I will try to get some thoughts back to you soon.

**Comment by jbbjarnason** [comment]
> Hi.
> 
> I have gotten interested in REPE as an RPC protocol. I gather there is excellent command support (request, response from client to server and back), but was thinking that not all communication goes one way, though it can be made to work like this, it is often not ideal to poll until the result is what you need from the client side.
> 
> For this an RPC protocol would sometimes have some event or state representation that a clients can subscribe to and at any given time receive an event from the server.
> 
> I can think of a few ways to make this work, like keep sending a command and once you get a response (the event) you "resubscribe" by sending another. Or host a server on the client too (though this will cause issues with port forwarding and the like) that the server sends these "events" as a request etc. 
> 
> Both are not ideal, so I'm hoping that I'm just missing this feature.
> 
> So is this already possible in REPE/Glaze? If so; how?

Seems like you are describing publish, subscribe messaging pattern, is that so or am I misinterpreting your description?

**Reply by SiebrenW** [reply]
Pretty much, yes. We call them events in my workplace, because we don't talk server-client, but interface server-proxy (if that makes sense) where the interface is divided in states, events and commands. Commands are your typical remote procedures, events something you get from the server as are states (these are just a stateful wrapper that is updated by the same principle of the event on change).

In detail that would likely mean publish-subscribe pattern (I'm not aware of an alternative). The means aren't that important to me, I just want a push event from the server.

**Comment by stephenberry** [comment]
@SiebrenW, I'm currently working on the REPE implementation in Glaze (#1044), making it thread safe for concurrent reading/writing. I will be using this with a websocket interface, which is similar to the publish-subscribe pattern. Once I've been able to implement and test the REPE implementation with websocket interface I should be able to give you a good answer about how to handle a publish-subscribe pattern, or an event pattern like you are describing. Just letting you know that progress is being made.

**Reply by SiebrenW** [reply]
Thanks for thinking about me.
Do let me know if I can help with anything. I have plenty of experience with boost asio (which is basically the same as lib asio that you use in glaze)
I would be happy to help.

**Reply by stephenberry** [reply]
Sweet, I'd love to get your help. I'd actually like to make the code work with both boost asio and the standalone asio. I'll reach out when I think I'm ready for a bit higher level development. I want to get a good implementation of the thread safety working before improving the server/client implementations.

**Reply by SiebrenW** [reply]
To be honest, I was actually semi-planning to throw an optional include of boost::asio over the fence once you were done with your REPE work.
I think an alias to boost::asio and asio in an ifdef would suffice, but "how hard can it be" are some famous last words.

**Reply by SiebrenW** [reply]
I just tested it on main (no fancy configuration yet). I changed the header asio.hpp to boost/asio.hpp and aliased boost::asio to asio and at least the current tests seem to work just fine.
Only issue is with the code I provided not too long ago of all things where boost asio wants to use a boost::system::error_code instead of an std::error_code.
Can be fixed with a simple #if defined(GLZ_USE_BOOST_ASIO) or something.

I could throw you a PR right now if you like. I am just not sure about how we want to configure this. I was thinking the above define and maybe check if asio.hpp is not in the includes to see if boost/asio.hpp was there.

It would look something like this:
```cpp
#if __has_include(<asio.hpp>) && !defined(GLZ_USE_BOOST_ASIO)
#include <asio.hpp>
#elif __has_include(<boost/asio.hpp>)
#ifndef GLZ_USING_BOOST_ASIO
#define GLZ_USING_BOOST_ASIO
#endif
#include <boost/asio.hpp>
namespace asio = boost::asio;
#else
static_assert(false, "standalone asio must be included to use glaze/ext/glaze_asio.hpp");
#endif
```
And in the init() function this:
```cpp
#if !defined(GLZ_USING_BOOST_ASIO)
         std::error_code ec{};
#else
         boost::system::error_code ec{};
#endif
```
That's all that was needed.

**Reply by stephenberry** [reply]
This is looking good, go ahead and submit a pull request.
